_posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md

```diff
@@ -0,0 +1,137 @@
+Almost!
+Since Elixir is based on immutable data, you could not have started as shown in the first JavaScript solution.
+You cannot simply mutate a data structure in Elixir!
+What you would have done is to create a new list with every value from the old one that you see fit.
+Besides that, there are some other oddities.
+The `|>` is called a pipe operator and passes the value as the first argument to the next function.
+While we're using the `fn` keyword to define an anonymous function as the `filter` expression and explicitly naming the variable `num`, we're shortening this for the `map` function and just say: "*This function uses the first argument (`&1`) it receives and doubles it*".
+
+Let's take the example from above again.
+In contrast to JavaScript, we can use some sugar in Elixir called a `for` comprehension in conjunction with a filter to loop over the enumerable and create a new one (with all values > 0 doubled).
+
+```elixir
+my_numbers = [-2, 3, 5, -34, 0, 32]
+result = for n <- my_numbers, n > 0, do: n * 2
+```
+
+Note that the last statement of every Elixir function is automatically returned creating a new list bound to `result`.
+What actually looks like an assignment `=` is actually a `match`.
+Trying to match `result = 4` afterwards will give you an error, since the compiler tries to match `[6, 10, 64] = 4`.
+This might look strange at first, yet makes more sense once you grasp concepts like pattern matching or destructuring.
+
+After looking at the above described language features, I cannot circumvent a few basic facts.
+Elixir runs on the [Erlang](https://www.erlang.org/) VM (BEAM) and gives you the great ecosystem companies like *Heroku* and *WhatsApp* use to build their products.
+It kind of relates to Erlang like Scala relates to Java (JVM).
+Likewise, you can import Erlang modules and use them right away.
+Elixir's syntax is concise and reminds of the Ruby programming language due to the fact that the creator [José Valim](https://github.com/josevalim) is a very active member of the ruby community.
+
+
+Ok, let's see some more Elixir code and compare it to JavaScript ES6.
+
+Elixir
+
+```elixir
+# Printing the sum of a list
+Enum.reduce([1, 2, 3, 5, 8], fn(item, acc) -> acc + item end) |> IO.puts
+
+# or without explicitely naming the parameters
+Enum.reduce([1, 2, 3, 5, 8], &(&1 + &2)) |> IO.puts
+```
+
+JS
+
+```js
+// Printing the sum of an array
+console.log([1, 2, 3, 5, 8].reduce((acc, item) => acc + item))
+```
+
+Elixir
+
+```elixir
+# And we all love fibonacci right?
+defmodule Fibonacci do
+    def calc (n) when n < 2, do: 1
+    def calc (n), do: calc(n - 2) + calc(n - 1)
+end
+```
+JS
+
+```js
+// We all do love fibonacci!
+function fibonacci (n) {
+   if (n < 2) return 1
+
+   return fibonacci(n - 2) + fibonacci(n - 1)
+}
+```
+
+### Enough of the platform
+
+I won't bother you anymore with advantages or language features Elixir provides nor try to persuade you that this is the right language to learn and use.
```

Comment:
At this point I'll stop listing language features Elixir provides since this post is only meant for sweetening. I will also not try to persuade you that [...]